File: Javascript/divide-conquer/assignment1/karatsuba.js
function karatsuba(strnum1, strnum2) {
  var num1 = BigInt(strnum1);
  var num2 = BigInt(strnum2);
  if (num1 < 10n || num2 < 10n) {
    return num1 * num2;
  }
  var strnum1len = (strnum1.toString()).length;
  var strnum2len = (strnum2.toString()).length;

  var len = strnum1len > strnum2len ? strnum1len : strnum2len
  var maxstrlen = BigInt(len) / 2n; 
  var base = 10n ** maxstrlen;

  var a = num1 / base;
  var b = num1 % base;
  var c = num2 / base;
  var d = num2 % base;

  var first = karatsuba(a, c);
  var second = karatsuba(b, d);
  var third = karatsuba(a + b, c + d);
  var fourth = karatsuba(a, d) + karatsuba(c, b)

  return BigInt(first) * (base ** 2n) + BigInt(second) + base * BigInt(fourth);

}

module.exports = karatsuba;

File: Python/divide-conquer/assignment1/karatsuba.py
import math

def karatsuba(strnum1, strnum2):
  num1 = int(strnum1)
  num2 = int(strnum2)

  if num1 < 10 or num2 < 10:
    return num1 * num2

  maxlen = max(len(str(num1)), len(str(num2))) // 2
  base = 10 ** maxlen

  def decompose(num):
    return num // base, num % base
  
  a, b = decompose(num1)
  c, d = decompose(num2)

  first = karatsuba(a ,c)
  second = karatsuba(b, d)
  third = karatsuba(a + b, c + d)

  return first * (base ** 2) + second + base * (third - first - second)